go/tools/builders/ar.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+type genFileInfo struct {
+	base       string       // The basename of the path
+	path       string       // The full path to the final file
+	expected   bool         // Whether the file is expected by the rules
+	created    bool         // Whether the file was created by protoc
+	from       *genFileInfo // The actual file protoc produced if not Path
+	unique     bool         // True if this base name is unique in expected results
+	ambiguious bool         // True if there were more than one possible outputs that matched this file
+}
+
+const (
+	// arHeader appears at the beginning of archives created by "ar" and
+	// "go tool pack" on all platforms.
+	arHeader = "!<arch>\n"
+
+	// entryLength is the size in bytes of the metadata preceding each file
+	// in an archive.
+	entryLength = 60
+)
+
+var zeroBytes = []byte("0                    ")
+
+type header struct {
+	NameRaw     [16]byte
+	ModTimeRaw  [12]byte
+	OwnerIdRaw  [6]byte
+	GroupIdRaw  [6]byte
+	FileModeRaw [8]byte
+	FileSizeRaw [10]byte
+	EndRaw      [2]byte
+}
+
+func (h *header) Name() string {
+	return strings.TrimRight(string(h.NameRaw[:]), " ")
+}
+
+func (h *header) Size() int64 {
```

Comment:
Nit: I don't think `Size`, `Next` or `Deterministic` need to be exported. I guess the fields need to be public though, since `binary` uses reflection.